rln-cli/src/examples/relay.rs

```diff
@@ -0,0 +1,298 @@
+        &mut self,
+        previous_proof: Vec<u8>,
+        current_proof: Vec<u8>,
+    ) -> Result<()> {
+        let mut proof1 = Cursor::new(previous_proof);
+        let mut proof2 = Cursor::new(current_proof);
+        let mut output = Cursor::new(Vec::new());
+
+        match self
+            .rln
+            .recover_id_secret(&mut proof1, &mut proof2, &mut output)
+        {
+            Ok(_) => {
+                let output_data = output.into_inner();
+                let (leaked_identity_secret_hash, _) = bytes_le_to_fr(&output_data);
+
+                if let Some((user_index, identity)) = self
+                    .local_identities
+                    .iter()
+                    .find(|(_, identity)| {
+                        identity.identity_secret_hash == leaked_identity_secret_hash
+                    })
+                    .map(|(index, identity)| (*index, identity))
+                {
+                    let real_identity_secret_hash = identity.identity_secret_hash;
+                    if leaked_identity_secret_hash != real_identity_secret_hash {
+                        Err(eyre!("identity secret hash mismatch {leaked_identity_secret_hash} != {real_identity_secret_hash}"))
+                    } else {
+                        println!("DUPLICATE message ID detected! Reveal identity secret hash: {leaked_identity_secret_hash}");
+                        self.local_identities.remove(&user_index);
+                        self.rln.delete_leaf(user_index)?;
+                        println!("User index {user_index} has been SLASHED");
+                        Ok(())
+                    }
+                } else {
+                    Err(eyre!(
+                        "user identity secret hash {leaked_identity_secret_hash} not found"
+                    ))
+                }
+            }
+            Err(err) => Err(eyre!("Failed to recover identity secret: {err}")),
+        }
+    }
+}
+
+fn main() -> Result<()> {
+    let mut rln_system = RLNSystem::new()?;
+    let rln_epoch = hash_to_field(b"epoch");
+    let rln_identifier = hash_to_field(b"rln-identifier");
+    let external_nullifier = poseidon_hash(&[rln_epoch, rln_identifier]);
+    println!("RLN Relay Example:");
+    println!("Message Limit: {MESSAGE_LIMIT}");
+    println!("----------------------------------");
+    println!();
+    show_commands();
+    loop {
+        print!("\n> ");
+        stdout().flush()?;
+        let mut input = String::new();
+        stdin().read_line(&mut input)?;
+        let trimmed = input.trim();
+        let args = std::iter::once("rln").chain(trimmed.split_whitespace());
+
+        match Cli::try_parse_from(args) {
+            Ok(cli) => match cli.command {
+                Commands::List => {
+                    rln_system.list_users();
+                }
+                Commands::Register => {
+                    rln_system.register_user()?;
+                }
+                Commands::Send {
```

Comment:
If I call this command twice with same message id it crashes

command: `send -u 0 -m 0 -s hi`
error: ```thread 'main' panicked at /Users/evb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ark-ff-0.4.2/src/fields/models/fp/mod.rs:907:42:
called `Option::unwrap()` on a `None` value```

Reply:
hm, I'll check it

Reply:
did you try to check through the RUST_BACKTRACE=1? 

So I found the problem - when you recover a secret, it actually assumes that we have two different signals for the same identifier and nullifiers. In the case where we send the same signal with the same parameters twice, there is no difference between the proofs and you divide by zero https://github.com/vacp2p/zerokit/blob/1930ca1610b85d59909fbbb6936ca042b6ebf1b8/rln/src/protocol.rs#L493

You need to add a check inside the ypur part that if the proof is the same, then you reject that message because it's duplicated and you can't even use it to recover the secret